### PR TITLE
Implemented homie 3.0.1 convention

### PR DESCRIPTION
**Pull request by marcus-garvey**
I didn't change the test cases. I'm sorry about it but I am to lazy :)

### CODE REVIEW

File: lib/homieDevice.js [hunk 1]
```diff
@@ -117,7 +117,7 @@ proto.setup = function(quiet) {
   })
 
   t.mqttClient.on('message', function (topic, message) {
-    t.onMessage(topic, message.toString());
```
Comment (lorenwest) [inline lib/homieDevice.js:118]: Why did you remove the toString() here? This will break existing clients because if a buffer comes in, it won't be converted for them. Did you need a feature that this prevented?

File: test/2-HomieNode.js [hunk 2]
```diff
@@ -11,13 +11,13 @@ describe("Homie Node", function() {
   describe("Instantiation", function() {
 
     var testDevice = new HomieDevice('homie-device-test');
-    var testNode1 = testDevice.node('test-node-1', 'test-node');
+    var testNode1 = testDevice.node('test-node-1', 'friendly Name','test-node');
```
Comment (lorenwest) [inline test/2-HomieNode.js:14]: If friendly name were added to the end of the list, would it make it compatible with earlier versions?

It's important to consider the impact of new features to existing implementations. Best case scenario, they can install the latest without any changes to their code - then migrate to new functionality as time permits.

If we introduce incompatibilities we'll have to bump the major version of this library. We may want or need to in order to support 3.x, but it's a huge win if we can bring existing users up to 3.x with little or no effort on their part.

**Review by lorenwest — COMMENTED** [review]
Thank you for your work in moving this to 3.0.1. I haven't had a chance to review the full specification, so a couple questions:

1) Do you believe this is the full set of changes necessary for 3.0.1?
2) Is there a way to minimize the migration of existing users to this version?
3) Tests are a pain in the a**, but they build trust in the library and show it's worthy of use. Before this goes live it will need coverage of all new functionality.
4) If the migration can't be made seamless, before committing we will need to include an upgrade manual. A list of minimum actions a user must take to use the new version. Ideally we can make this un-necessary.

If you want help with some of the above, we can ask for collaboration and I can schedule some time. It's something I've planned on doing for a while.

File: lib/homieDevice.js (hunk 1 above)
Comment (marcus-garvey) [reply lib/homieDevice.js:118]: One testcase failed with a null pointer exception with the toString method. It was only a test and I forgot to add a check

File: test/2-HomieNode.js (hunk 2 above)
Comment (marcus-garvey) [reply test/2-HomieNode.js:14]: The friendly name in the homie 3.0.1 convention is not optional, so I decided to change this so it better match with the homie esp8266 API.

Sure I would prefer small changes too, but the specification changed alot since version 2. 
The most mqtt messages and node property definitions are completely different.

**Comment by marcus-garvey** [comment]
1. Yes I think every feature is implemented, I will test it against openhab in the next days.
Some more helpers for $stats and $implementation could be added but these are optional.

2. Most of the changes have no impact on the interface. Only the array support changed a lot.
The 3.0.1 specification added more features to describe an property better now. Here we have some new methods, you can ignore these methods but in this case the MQTT messages are not complete and the device/node is not compatible to the specification. So a migration of the API syntax is easy but you will get a device/node which will not comply to version 2 or version 3. You have to use the new methods to create correct version 3 node.
Maybe it would be better to create a new lib homie-device-3 ? So the version 2 lib can still be used in npm?
The current API syntax matches to the current esp8266 homie implementation. 

3. Yes sure, I will add some additional test cases if I could find some time

**Comment by lorenwest** [comment]
Managing two libraries seems overkill. Probably easiest to bump the version of this library up to match the Homie convention version. That gives us the ability to fix bugs in the pre-3.x version without creating a different library.  We can keep the Major.Minor in synch with the spec, and bump the .Patch version as we release bugfixes of this library.

**Comment by puschie286** [comment]
any news on this ?

**Comment by lorenwest** [comment]
Test coverage was pretty solid for the previous version. I'd like to see test coverage for the new features before merging. @puschie286 Would you like to make that contribution?
